File: src/main/java/dev/syoritohatsuki/duckyupdater/DuckyUpdater.java
package dev.syoritohatsuki.duckyupdater;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.mojang.logging.LogUtils;
import dev.syoritohatsuki.duckyupdater.dto.UpdateData;
import net.fabricmc.loader.api.FabricLoader;
import net.minecraft.util.Pair;
import org.slf4j.Logger;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.HashMap;
import java.util.Map;

public class DuckyUpdater {

    public static final Logger logger = LogUtils.getLogger();

    public static Map<Pair<String, String>, UpdateData> requestUpdates() {

        final var updateMap = new HashMap<Pair<String, String>, UpdateData>();

        FabricLoader.getInstance().getAllMods().forEach(modContainer -> {
            if (modContainer.getMetadata().getCustomValue("duckyupdater") == null) return;
            try {
                var url = StringUtil.buildUrl(modContainer);
                if (url == null) return;

                var json = new Gson().fromJson(
                        HttpClient.newHttpClient().send(HttpRequest.newBuilder()
                                .uri(URI.create(url))
                                .setHeader("User-Agent", StringUtil.userAgent(modContainer))
                                .GET()
                                .build(), HttpResponse.BodyHandlers.ofString()).body(), JsonElement.class
                ).getAsJsonArray().get(0).getAsJsonObject();

                var remoteVersion = json.get("version_number").getAsString();
                if (remoteVersion.equals(modContainer.getMetadata().getVersion().getFriendlyString())) return;

                var updateData = new UpdateData(remoteVersion, json.get("changelog").getAsString(), json.get("version_type").getAsString(), json.get("files").getAsJsonArray().get(0).getAsJsonObject().get("url").getAsString());

                var type = "release";
                var typeObject = modContainer.getMetadata().getCustomValue("duckyupdater").getAsObject().get("type");
                if (typeObject != null) type = typeObject.getAsString();

                if (!updateData.type().equals(type)) return;

                updateMap.put(new Pair<>(modContainer.getMetadata().getName(), modContainer.getMetadata().getVersion().getFriendlyString()), updateData);
            } catch (Exception e) {
                logger.warn("Can't get update for " + modContainer.getMetadata().getId(), e);
            }
        });
        return updateMap;
    }
}

File: src/main/kotlin/dev/syoritohatsuki/duckyupdater/DuckyUpdater.kt
package dev.syoritohatsuki.duckyupdater

import com.google.gson.Gson
import com.google.gson.JsonElement
import com.mojang.logging.LogUtils
import net.fabricmc.loader.api.FabricLoader
import net.minecraft.SharedConstants
import org.slf4j.Logger
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse

object DuckyUpdater {

    val logger: Logger = LogUtils.getLogger()

    fun requestUpdates(): Map<Pair<String, String>, UpdateData> =
        mutableMapOf<Pair<String, String>, UpdateData>().apply {
            FabricLoader.getInstance().allMods.forEach { container ->
                if (container.metadata.getCustomValue("duckyupdater") == null) return@forEach
                try {
                    (Gson().fromJson(
                        HttpClient.newHttpClient().send(
                            HttpRequest.newBuilder()
                                .uri(
                                    URI.create(
                                        buildString {
                                            append("https://api.modrinth.com/v2/project/")
                                            append(
                                                (container.metadata.getCustomValue("duckyupdater")
                                                    ?: return@forEach).asObject.get("modrinthId").asString
                                            )
                                            append("/version?loaders=[%22fabric%22]&game_versions=[%22")
                                            append(SharedConstants.getGameVersion().name)
                                            append("%22]&featured=")
                                            append(
                                                container.metadata.getCustomValue("duckyupdater").asObject.get("featured")?.asBoolean
                                                    ?: false
                                            )
                                        }
                                    )
                                )
                                .setHeader(
                                    "User-Agent",
                                    "syorito-hatsuki/ducky-updater-lib/${
                                        FabricLoader.getInstance().getModContainer("ducky-updater-lib")
                                            .orElseThrow().metadata.version.friendlyString
                                    } (syorito-hatsuki.dev)"
                                )
                                .GET()
                                .build(), HttpResponse.BodyHandlers.ofString()
                        ).body(), JsonElement::class.java
                    ).asJsonArray[0]?.asJsonObject ?: return@forEach).let { json ->

                        if (json.get("version_number").asString == container.metadata.version.friendlyString) return@forEach

                        UpdateData(
                            json.get("version_number").asString,
                            json.get("changelog").asString,
                            json.get("version_type").asString,
                            json.get("files").asJsonArray[0].asJsonObject.get("url").asString
                        ).also {
                            if (it.type == (container.metadata.getCustomValue("duckyupdater").asObject.get("type")?.asString
                                    ?: "release")
                            ) this[Pair(
                                container.metadata.name,
                                container.metadata.version.friendlyString
                            )] = it
                        }
                    }
                } catch (e: Exception) {
                    logger.warn("Can't get update for ${container.metadata.id}", e)
                }
            }
        }

    fun match(oldVersion: CharArray, newVersion: CharArray): String {
        val result = StringBuilder()
        try {
            var index = 0
            while (oldVersion[index] == newVersion[index]) {
                result.append(oldVersion[index])
                index++
            }
        } catch (_: ArrayIndexOutOfBoundsException) {
        }
        return result.toString()
    }

    class UpdateData(
        val remoteVersion: String,
        val changelog: String,
        val type: String,
        val fileUrl: String
    )
}